### ISSUE TITLE
Need example for combining Unicode audio tokens with text prompts for Spire LM + gibberish or blank output from inference

### ISSUE THREAD

**Issue by cdkmd1234** [utter-project/SpireLM#4]
Hi,

1. Need an example for combining Unicode speech tokens with text
I have Unicode-style speech tokens obtained via the Labeler (from HuBERT + k-means), like:

\U000f025e\U000f098e\U000f11ae\U000f03c5...
I would like to input these tokens alongside a textual question (e.g., a QA task or translation prompt) into the SpireBase model.

Is there an official example or format for structuring a prompt combining audio tokens and text?

2. Inference returns blank or corrupted outputs
Here is the code snippet I’m currently using:

model = AutoModelForCausalLM.from_pretrained("utter-project/SpireBase", torch_dtype=torch.float16, device_map="auto")
tokenizer = AutoTokenizer.from_pretrained("utter-project/SpireBase")
tokenizer.pad_token = tokenizer.eos_token
    
prompts = f"""Translate the following text from English to German.\nEnglish: {unicode}\nGerman:""" 
inputs = tokenizer(prompts, return_token_type_ids=False, return_tensors="pt", padding=True).to(model.device)
with torch.no_grad():
    outputs = model.generate(**inputs, max_new_tokens=100)
for i in outputs:   
    generated_text = tokenizer.decode(i, skip_special_tokens=True)

Problem:
The model returns the following output — a long sequence of mostly newlines and backslashes, with no meaningful text:

\n\\\\\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n
This happens consistently regardless of question content or audio input.

Questions / Requests
Is the prompt formatting incorrect, leading to the invalid outputs?

Are there special decoding or tokenization steps needed for Unicode audio tokens?

Could you provide a minimal working example of how to:

encode speech tokens + text,

pass them to SpireBase, and

correctly decode the result?

Thank you in advance for your help 

Best regards,
Liam

**Comment by bpopeters** [comment]
Hi Liam,

Sorry for the long time in replying, I've been in a mix of deadlines and vacation. I'll add some examples to the repo soon. But in the meantime, here are some pointers:

1. I wouldn't recommend using `SpireBase`, as it hasn't been instruction-tuned and is likely to perform poorly in most instances. `SpireFull` should work better in basically all instances.
2. In the example above, the model is loaded as `torch.float16`, aka fp16. We didn't do any of the training or inference with fp16, and in general it tends to be somewhat unstable unless you take specific precautions. I would recommend either `torch.float` (fp32) or `torch.bfloat16` (bf16) (the inference in the paper was performed using bf16).
3. Now, the root of the issue: `SpireFull` expects the prompts to use a chat template. So let's say, for example that use the following ASR prompt (which is the one we used in our experiments):

```
prompt = Speech: 󰀦󰡂󰱒󰂋󰤗󱅮\n English:
```

In order to turn this into chat template format, you can do the following:

```
messages = [{"role": "user", "content": prompt}]
formatted_prompt = tokenizer.apply_chat_template(messages, tokenize=False, add_generation_prompt=True)
```

You can then tokenize `formatted_prompt` and pass it to `model.generate`.

Hope that helps,
Ben

**Comment by saikoneru** [comment]
I attached a sample inference code (German is hard-coded but can be adapted) if anybody needs. (Also thanks authors for their work) 

```
 import os
 from transformers import AutoModelForCausalLM, AutoTokenizer, GenerationConfig
 import argparse
 from spire.hubert_labeler import HubertLabeler as Labeler

 def read_file(file_path):
     with open(file_path, 'r') as f:
         lines = f.readlines()
     lines = [x.rstrip() for x in lines]
     return lines

 def write_file(file_path, lines):
     with open(file_path, 'w') as f:
         for line in lines:
             f.write(str(line) + '\n')

 def divide_chunks(l, n):
          for i in range(0, len(l), n):
              yield l[i:i + n]

 def parse_args():
     parser = argparse.ArgumentParser(description="Evaluate Tower Spire")

     parser.add_argument('--tgt_lang', type=str, required=True, help='Target language')
     parser.add_argument('--input_fp', type=str, required=True, help='Input Audio Filepaths')
     parser.add_argument('--output_fp', type=str, required=True, help='Output file')
     return parser.parse_args()

 def main():
     args = parse_args()
     audio_fps = read_file(params.input_fp)

     tgt_langs = [args.tgt_lang] * len(audio_fps)

     model_path = "utter-project/SpireFull"

     tokenizer = AutoTokenizer.from_pretrained(model_path, trust_remote_code=True)
     model =  AutoModelForCausalLM.from_pretrained(model_path, torch_dtype=torch.bfloat16, device_map="auto", trust_remote_code=True).eval()

     generation_config = GenerationConfig.from_pretrained(model_path)


     batch_size = 1
     audio_batches = divide_chunks(audio_fps, batch_size)
     lang_batches = divide_chunks(tgt_langs, batch_size)

     labeler = Labeler("facebook/hubert-large-ll60k", "kmeans_model")


     outputs = []

     with torch.inference_mode():

         for audio_batch in audio_batches:

             tgt_lang = tgt_langs[idx]
             curr_langs = next(lang_batches)
             #curr_asr_hyps = next(asr_batches)

             prompts = []
             audios = []

             for idx in range(len(audio_batch)):
                 speech_tokens = labeler.label_wav(audio_batch[idx])[0]
                 # For self-cascade, you can use a prompt like this
                 #prompt = "Speech: " + speech_tokens + "\n" + "English: " + curr_asr_hyps[idx] + "\nGerman" + ":"
                 prompt = "Speech: " + speech_tokens + "\n" + "German: "
                 messages = [{"role": "user", "content": prompt}]
                 formatted_prompt = tokenizer.apply_chat_template(messages, tokenize=False,add_generation_prompt=True)
                 prompts.append(formatted_prompt)


             inputs = tokenizer(prompts, return_tensors='pt', padding=True).to(model.device)

             generate_ids = model.generate(
                 **inputs,
                 max_new_tokens=256,
                 num_beams=5,
                 eos_token_id= model.config.eos_token_id,
                 generation_config=generation_config,
             )

             generate_ids = generate_ids[:, inputs['input_ids'].shape[1]:]

             responses = tokenizer.batch_decode(
                 generate_ids, skip_special_tokens=True, clean_up_tokenization_spaces=False
             )

             outputs.extend(responses)
             print(f'>>> Response\n{responses}')


     write_file(args.output_fp, outputs)


 if __name__ == "__main__":
     main()

```

**Comment by bpopeters** [comment]
Hi Sai, thanks for the cleaner illustration than in `inference.py`. How well does beam search work?

**Comment by saikoneru** [comment]
Hi Ben, i did not compare to greedy search but i have a comet score around 85 for FLEURS EN-DE with beam 5 self-cascade
